### PR TITLE
[MERGE] (test_)mail(_full): cleanup composer and post tests, add portal / rating

### PR DESCRIPTION
**Pull request by tde-banana-odoo**
PURPOSE

Cleanup and improve test coverage in test_mail(_full), notably about composer,
message_post API (and its sub-methods) and attachments management.

SPECIFICATIONS

Improve tooling of mail tests. Use SMTP mockup class available in base tests
in mail tests to have a better and simplified mock tool when dealing with
mail tests. Remove old fashioned code.

Add tests related to mail failure management and its effect on mail.mail
and related mail.notification records.

Cleanup and improve composer-related tests. As composer will soon be improved
better ensure we don't break flows.

Cleanup and improve post-related tests. As composer and posting methods will
soon be improved better ensure we don't break flows.

Add some first tests about rating and rating mixin:
  * rating preparation: token, information from the record to rate;
  * rating apply: posting message, author;

See sub commits for more details.

LINKS

Task-2673913 (TestMail: Cleanup and improve test coverage)
Task-2704390 (TestMailFull: add first rating tests)

Prepares Task-2088884 (Mail: Composer Onchange to Editable Computed Stored)
Prepares Task-2710804 (Mail: Clean MailThread API)
Prepares Task-2684479 (Mail: Better send errors management)
Prepares Task-36879 (Mail: Support MultiCompany Aliases)

### CODE REVIEW

**Comment by tde-banana-odoo** [comment]
@std-odoo would gladly have your word on this one (followup of a validation you did, to merge smtplib case and mockemail). Plus some other tests update, because why not.

File: addons/test_mail/tests/test_mail_mail.py [hunk 1]
```diff
@@ -65,6 +80,380 @@ def test_mail_mail_return_path(self):
...
+        }])
+        notification = self.env['mail.notification'].create({
+            'is_read': False,
+            'mail_mail_id': mail.id,
+            'mail_message_id': message.id,
+            'notification_type': 'email',
+            'res_partner_id': self.partner_employee.id,  # not really used for matching except multi-recipients
+        })
+
+        def _reset_data():
```
Comment (std-odoo) [inline addons/test_mail/tests/test_mail_mail.py:103]: isn't it better to split this test function ? (so you do not need `_reset_data`, and you have one test per exception)

File: addons/test_mail/tests/test_mail_mail.py [hunk 2]
```diff
@@ -65,6 +80,380 @@ def test_mail_mail_return_path(self):
...
+        self.assertEqual(notification.notification_status, 'exception')
+
+        # MailServer.send_email(): _prepare_email_message: Malformed 'Return-Path' or 'From' address
+        self.env['ir.config_parameter'].set_param('mail.catchall.domain', '')
+        _reset_data()
+        mail.write({'email_from': 'robert'})
+        with self.mock_mail_gateway(), mute_logger('odoo.addons.mail.models.mail_mail'):
+            mail.send(raise_exception=False)
+        self.assertEqual(self._mails[0]['email_from'], 'robert')
+        self.assertIn('Malformed \'Return-Path\' or \'From\' address: robert', mail.failure_reason)
```
Comment (std-odoo) [inline addons/test_mail/tests/test_mail_mail.py:159]:
```suggestion
        self.assertIn("Malformed 'Return-Path' or 'From' address: robert", mail.failure_reason)
```

File: addons/test_mail/tests/test_message_post.py [hunk 3]
```diff
@@ -347,58 +574,93 @@ def test_post_attachments(self):
...
+    def test_post_internal(self):
+        test_record = self.env['mail.test.simple'].browse(self.test_record.ids)
+
+        test_record.message_subscribe([self.user_admin.partner_id.id])
+        with self.mock_mail_gateway():
+            msg = test_record.message_post(
+                body='My Body',
+                message_type='comment',
+                subject='My Subject',
+                subtype_xmlid='mail.mt_note'
```
Comment (std-odoo) [inline addons/test_mail/tests/test_message_post.py:635]:
```suggestion
                subtype_xmlid='mail.mt_note',
```

File: addons/test_mail/tests/test_performance.py [hunk 4]
```diff
@@ -43,11 +45,18 @@ def setUpClass(cls):
             'mobile': '0456123456',
         })
 
+        cls.test_attachments_vals = cls._generate_attachments_data(
+            cls, 3,
+            res_model='mail.compose.message',
+            res_id=0
+        )
+
```
Comment (std-odoo) [inline addons/test_mail/tests/test_performance.py:53]:
```suggestion
        cls.test_attachments_vals = cls._generate_attachments_data(
            cls, 3, res_model='mail.compose.message', res_id=0)
```

File: addons/test_mail/tests/test_performance.py [hunk 5]
```diff
@@ -296,7 +321,38 @@ def test_adv_activity_mixin(self):
...
+        attachments = self.env['ir.attachment'].create([
+            dict(values,
+                 res_model='mail.activity',
+                 res_id=0)
+            for values in self.test_attachments_vals
+        ])
+
+        with self.assertQueryCount(__system__=8, employee=8):
+            activity = record.action_start('Test Start')
+            #read activity_type to normalize cache between enterprise and community
```
Comment (std-odoo) [inline addons/test_mail/tests/test_performance.py:337]: (not pep compliant, there's also other occurrence in this pr)
```suggestion
            # read activity_type to normalize cache between enterprise and community
```

File: addons/test_mail_full/tests/test_rating.py [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo.addons.test_mail_full.tests.common import TestMailFullCommon, TestMailFullRecipients
+from odoo.tests.common import users
+from odoo.tools import mute_logger
+from odoo.tests import tagged
+
+
+@tagged('rating')
+class TestRatingCommon(TestMailFullCommon, TestMailFullRecipients):
```
Comment (std-odoo) [inline addons/test_mail_full/tests/test_rating.py:11]: can you merge it with `TestRatingFlow` ? (look like it's only used in this file)

File: addons/test_mail_full/tests/test_rating.py [hunk 7]
```diff
@@ -0,0 +1,78 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo.addons.test_mail_full.tests.common import TestMailFullCommon, TestMailFullRecipients
+from odoo.tests.common import users
+from odoo.tools import mute_logger
+from odoo.tests import tagged
```
Comment (std-odoo) [inline addons/test_mail_full/tests/test_rating.py:7]: a grand sage would ask me to put it before
```
from odoo.tests.common import users
```

File: addons/test_mail/tests/test_performance.py [hunk 8]
```diff
@@ -66,6 +75,22 @@ def _flush_tracking(self):
         self.env['base'].flush()
         self.cr.flush()
 
+    def _generate_attachments_data(self, count, res_model=None, res_id=None, attach_values=None):
+        # attachment visibility depends on what they are attached to
+        attach_values = attach_values if attach_values else {}
```
Comment (std-odoo) [inline addons/test_mail/tests/test_performance.py:80]:
```suggestion
        attach_values = attach_values or {}
```

File: addons/test_mail/tests/test_message_post.py [hunk 9]
```diff
@@ -225,115 +260,307 @@ def test_notify_recipients_internals(self):
...
             ('List2', b'My second attachment')
         ]
-        _attach_1 = self.env['ir.attachment'].with_user(self.user_employee).create({
+        _attach_1 = self.env['ir.attachment'].create({
             'name': 'Attach1',
             'datas': 'bWlncmF0aW9uIHRlc3Q=',
-            'res_model': 'mail.compose.message', 'res_id': 0})
-        _attach_2 = self.env['ir.attachment'].with_user(self.user_employee).create({
+            'res_model': 'mail.compose.message',
+            'res_id': 0
```
Comment (std-odoo) [inline addons/test_mail/tests/test_message_post.py:545]: (check also other occurrence in this pr)
```suggestion
            'res_id': 0,
```

File: addons/test_mail/tests/test_message_post.py [hunk 10]
```diff
@@ -225,115 +260,307 @@ def test_notify_recipients_internals(self):
...
+        with self.mock_mail_gateway(mail_unlink_sent=True):
+            msg = test_record.message_post(
+                body='Test', subject='Test',
+                message_type='comment', subtype_xmlid='mail.mt_comment',
+                partner_ids=[self.partner_1.id, self.partner_2.id],
+                mail_auto_delete=False
+            )
 
         # notifications emails should not have been deleted: one for customers, one for user
         self.assertEqual(len(self.env['mail.mail'].sudo().search([('mail_message_id', '=', msg.id)])), 2)
```
Comment (std-odoo) [inline addons/test_mail/tests/test_message_post.py:527]: while you are at it
```suggestion
        self.assertEqual(self.env['mail.mail'].sudo().search_count([('mail_message_id', '=', msg.id)]), 2)
```

File: addons/test_mail/tests/test_message_post.py [hunk 11]
```diff
@@ -225,115 +260,307 @@ def test_notify_recipients_internals(self):
...
+        self.assertEqual(len(self.env['mail.mail'].sudo().search([('mail_message_id', '=', new_message.id)])), 0)
+
+        with self.assertSinglePostNotifications([
+                {'partner': self.partner_1, 'type': 'email'},
+                {'partner': self.partner_portal, 'type': 'email'}], {'content': 'ToPortal'}):
+            test_record.message_post(
+                body='ToPortal',
+                message_type='comment',
+                subtype_xmlid='mail.mt_comment',
+                partner_ids=[self.partner_portal.id])
```
Comment (std-odoo) [inline addons/test_mail/tests/test_message_post.py:496]:
```suggestion
                partner_ids=self.partner_portal.ids)
```

File: addons/test_mail/tests/test_message_post.py [hunk 12]
```diff
@@ -45,8 +64,19 @@ def setUpClass(cls):
         })
         cls.user_admin.write({'notification_type': 'email'})
 
+    def setUp(self):
+        super(TestMessagePostCommon, self).setUp()
+        # send tracking and messages to avoid nondeterministic tests
+        self.flush_tracking()
+        # patch registry to simulate a ready environment
+        self.patch(self.env.registry, 'ready', True)
```
Comment (std-odoo) [inline addons/test_mail/tests/test_message_post.py:72]: why is it needed ?

File: addons/test_mail/tests/test_message_post.py [hunk 13]
```diff
@@ -19,12 +20,30 @@
...
+        cls.user_portal = cls._create_portal_user()
+        cls.partner_portal = cls.user_portal.partner_id
+
+        # another standard employee to test follow and notifications between two
+        # users (and not admin / user)
+        cls.user_employee_2 = mail_new_test_user(
+            cls.env, login='employee2',
+            groups='base.group_user',
+            company_id=cls.company_admin.id,
+            # email='"Eglantine Employee2" <eglantine@example.com>',
```
Comment (std-odoo) [inline addons/test_mail/tests/test_message_post.py:39]:
```suggestion
```

File: addons/test_mail/tests/test_mail_mail.py [hunk 14]
```diff
@@ -65,6 +80,380 @@ def test_mail_mail_return_path(self):
...
+            _reset_data()
+            with mute_logger('odoo.addons.mail.models.mail_mail'):
+                mail.send(raise_exception=False)
+            self.assertEqual(mail.failure_reason, 'Unexpected issue')
+            self.assertFalse(mail.failure_type, 'Mail: unlogged failure type to fix')
+            self.assertEqual(mail.state, 'exception')
+            self.assertEqual(notification.failure_type, 'unknown', 'Mail: generic failure type')
+            self.assertEqual(notification.notification_status, 'exception')
+
+            self.send_email_mocked.side_effect = _send_current
```
Comment (std-odoo) [inline addons/test_mail/tests/test_mail_mail.py:455]: those tests look very redundant, can't make a list of Exception / state / failure_type ?

**Review by std-odoo — COMMENTED** [review]
@tde-banana-odoo Hi :) 

First validation pass :)

File: addons/test_mail/tests/test_mail_mail.py [hunk 15]
```diff
@@ -65,6 +80,380 @@ def test_mail_mail_return_path(self):
...
+            _connect_current = self.connect_mocked.side_effect
+
+            for error, msg in [
+                    (smtplib.SMTPServerDisconnected('SMTPServerDisconnected'), 'SMTPServerDisconnected'),
+                    (smtplib.SMTPResponseException('code', 'SMTPResponseException'), 'code\nSMTPResponseException'),
+                    (smtplib.SMTPNotSupportedError('SMTPNotSupportedError'), 'SMTPNotSupportedError'),
+                    (smtplib.SMTPException('SMTPException'), 'SMTPException'),
+                    (SSLError('SSLError'), 'SSLError'),
+                    (gaierror('gaierror'), 'gaierror'),
+                    (timeout('timeout'), 'timeout')]:
```
[the remark this answers is not in the document]
Comment (tde-banana-odoo) [reply addons/test_mail/tests/test_mail_mail.py:413]: I globally looked at what can be raised, just if someday we want to be more precise in our failure messages. I agree gaierror is maybe not the clearest one but hey :D .

File: addons/test_mail/tests/test_message_post.py [hunk 16]
```diff
@@ -347,58 +574,93 @@ def test_post_attachments(self):
...
+
+        test_record.message_subscribe([self.user_admin.partner_id.id])
+        with self.mock_mail_gateway():
+            msg = test_record.message_post(
+                body='My Body',
+                message_type='comment',
+                subject='My Subject',
+                subtype_xmlid='mail.mt_note'
+            )
+        self.assertFalse(msg.is_internal)  # tde strange
```
[the remark this answers is not in the document]
Comment (tde-banana-odoo) [reply addons/test_mail/tests/test_message_post.py]: It is linked to an internal subtype, but not "internal". But ACLs support it, I guess that's ok ^^ .

[the remark this answers is not in the document]
Comment (tde-banana-odoo) [reply addons/test_mail/tests/test_message_post.py]: Added a comment.

**Review by tde-banana-odoo — COMMENTED** [review]
Post-review comments !

**Comment by tde-banana-odoo** [comment]
@robodoo override=ci/security

Reason: "OpenSSL" is indeed present, just to import and simulate an error in email sending.

File: addons/test_mail/tests/test_mail_composer.py [hunk 17]
```diff
@@ -656,10 +894,73 @@ def setUpClass(cls):
...
+                                  default_template_id=self.template.id)
+        ))
+        composer = composer_form.save()
+        composer.auto_delete = False
+        with self.mock_mail_gateway(mail_unlink_sent=True), self.mock_mail_app():
+            composer._action_send_mail()
+
+        self.assertEqual(len(self._mails), 2, 'Should have sent 1 email per record')
+        self.assertEqual(len(self._new_mails), 2, 'Should have created 1 mail.mail per record')
+        # self.assertEqual(self._new_mails.exists(), self._new_mails, 'Should not have deleted mail.mail records')
```
Comment (std-odoo) [inline addons/test_mail/tests/test_mail_composer.py:933]:
```suggestion
```

File: addons/test_mail/tests/test_mail_composer.py [hunk 18]
```diff
@@ -656,10 +894,73 @@ def setUpClass(cls):
...
+            self._get_web_context(self.test_records, add_web=True,
+                                  default_template_id=self.template.id)
+        ))
+        composer = composer_form.save()
+        with self.mock_mail_gateway(mail_unlink_sent=True), self.mock_mail_app():
+            composer._action_send_mail()
+
+        self.assertEqual(len(self._mails), 2, 'Should have sent 1 email per record')
+        self.assertEqual(len(self._new_mails), 2, 'Should have created 1 mail.mail per record')
+        self.assertEqual(self._new_mails.exists(), self.env['mail.mail'], 'Should have deleted mail.mail records')
```
Comment (std-odoo) [inline addons/test_mail/tests/test_mail_composer.py:917]: can you do this instead ?
```suggestion
        self.assertFalse(self._new_mails.exists(), 'Should have deleted mail.mail records')
```

File: addons/test_mail/tests/test_mail_composer.py [hunk 19]
```diff
@@ -656,10 +894,73 @@ def setUpClass(cls):
...
+        ))
+        composer = composer_form.save()
+        composer.auto_delete = False
+        with self.mock_mail_gateway(mail_unlink_sent=True), self.mock_mail_app():
+            composer._action_send_mail()
+
+        self.assertEqual(len(self._mails), 2, 'Should have sent 1 email per record')
+        self.assertEqual(len(self._new_mails), 2, 'Should have created 1 mail.mail per record')
+        # self.assertEqual(self._new_mails.exists(), self._new_mails, 'Should not have deleted mail.mail records')
+        self.assertEqual(self._new_mails.exists(), self.env['mail.mail'], 'Template is forced over composer value, which is not correct')
```
Comment (std-odoo) [inline addons/test_mail/tests/test_mail_composer.py:934]:
```suggestion
        self.assertFalse(self._new_mails.exists(), 'Template is forced over composer value, which is not correct')
```

File: addons/test_mail/tests/test_mail_composer.py [hunk 20]
```diff
@@ -656,10 +894,73 @@ def setUpClass(cls):
...
+        ))
+        composer = composer_form.save()
+        composer.auto_delete_message = True
+        with self.mock_mail_gateway(mail_unlink_sent=True), self.mock_mail_app():
+            composer._action_send_mail()
+
+        # global outgoing
+        self.assertEqual(len(self._mails), 2, 'Should have sent 1 email per record')
+        self.assertEqual(len(self._new_mails), 2, 'Should have created 1 mail.mail per record')
+        self.assertEqual(self._new_mails.exists(), self.env['mail.mail'], 'Should have deleted mail.mail records')
```
Comment (std-odoo) [inline addons/test_mail/tests/test_mail_composer.py:951]: same here, assertFalse or maybe `assertEqual(len(...))` ?

File: addons/test_mail/tests/test_mail_composer.py [hunk 21]
```diff
@@ -656,10 +894,73 @@ def setUpClass(cls):
...
+        self.assertEqual(len(self._new_msgs), 2, 'Should have created 1 mail.mail per record')
+        self.assertEqual(self._new_msgs.exists(), self.env['mail.message'], 'Should have deleted mail.message records')
```
Comment (std-odoo) [inline addons/test_mail/tests/test_mail_composer.py:953]: same here (to be consistent with above)

File: addons/test_mail/tests/test_mail_composer.py [hunk 22]
```diff
@@ -841,7 +1090,9 @@ def test_mail_composer_wtpl_recipients(self):
 
         for record in self.test_records:
             # template is sent directly using customer field
-            self.assertSentEmail(self.partner_employee, record.customer_id)
+            self.assertSentEmail(self.partner_employee_2.email_formatted, record.customer_id,
+                                 author_id=self.partner_employee
+                                )
```
Comment (std-odoo) [inline addons/test_mail/tests/test_mail_composer.py:1095]:
```suggestion
                                 author_id=self.partner_employee)
```

File: addons/test_mail/tests/test_mail_mail.py [hunk 23]
```diff
@@ -44,6 +74,18 @@ def test_mail_message_notify_from_mail_mail(self):
         self.assertSentEmail(mail.env.user.partner_id, ['test@example.com'])
         self.assertEqual(len(self._mails), 1)
 
+        mail = self.env['mail.mail'].sudo().create({
+            'body_html': '<p>Test</p>',
+            'email_to': 'test@example.com',
+            'recipient_ids': [(4, self.user_employee.partner_id.id)]
```
Comment (std-odoo) [inline addons/test_mail/tests/test_mail_mail.py:80]:
```suggestion
            'recipient_ids': [(4, self.user_employee.partner_id.id)],
```

**Review by std-odoo — COMMENTED** [review]
@tde-banana-odoo Hi,

Some more minor comments, then it should be good for me :)

File: addons/test_mail/tests/test_mail_composer.py [hunk 24]
```diff
@@ -108,15 +104,28 @@ class TestComposerForm(TestMailComposer):
...
-            composer_form.subject, 'Re: %s' % self.test_record.name,
-            'MailComposer: comment mode should have default subject Re: record_name')
+        composer_form = Form(self.env['mail.compose.message'].with_context(
+            self._get_web_context(self.test_record, add_web=True)
+        ))
+        # auto_delete not displayed currently in view
+        # self.assertFalse(composer_form.auto_delete)
+        # self.assertFalse(composer_form.auto_delete_message)
+        # author_id not displayed currently in view
+        # self.assertEqual(composer_form.author_id, self.env.user.partner_id)
```
[the remark this answers is not in the document]
Comment (tde-banana-odoo) [reply addons/test_mail/tests/test_mail_composer.py:114]: Yep. In task about removing the composer onchange I will add the fields as invisible and correctly support them with computed editable fields :) . That way it makes the diff clearer: from unsupported to officially supported.

File: addons/test_mail/tests/test_mail_composer.py [hunk 25]
```diff
@@ -171,28 +180,76 @@ def test_mail_composer_comment_attachments(self):
...
-        self.assertFalse(composer_form.subject, 'MailComposer: mass mode should have void default subject if no template')
+    def test_mail_composer_comment_wtpl(self):
+        composer_form = Form(self.env['mail.compose.message'].with_context(
+            self._get_web_context(self.test_record, add_web=True, default_template_id=self.template.id)
+        ))
+        # auto_delete not displayed currently in view
+        # self.assertTrue(composer_form.auto_delete)
+        # self.assertFalse(composer_form.auto_delete_message)
+        # author_id not displayed currently in view
+        # self.assertEqual(composer_form.author_id, self.env.user.partner_id)
```
[the remark this answers is not in the document]
Comment (tde-banana-odoo) [reply addons/test_mail/tests/test_mail_composer.py:191]: Same here.

File: addons/test_mail/tests/test_mail_composer.py [hunk 26]
```diff
@@ -171,28 +180,76 @@ def test_mail_composer_comment_attachments(self):
...
+    @users('employee')
+    def test_mail_composer_mass(self):
+        composer_form = Form(self.env['mail.compose.message'].with_context(
+            self._get_web_context(self.test_records, add_web=True)
+        ))
+        # auto_delete not displayed currently in view
+        # self.assertFalse(composer_form.auto_delete)
+        # self.assertFalse(composer_form.auto_delete_message)
+        # author_id not displayed currently in view
+        # self.assertEqual(composer_form.author_id, self.env.user.partner_id)
```
[the remark this answers is not in the document]
Comment (tde-banana-odoo) [reply addons/test_mail/tests/test_mail_composer.py:214]: Same answer :D

File: addons/test_mail/tests/test_mail_composer.py (hunk 18 above)
Comment (tde-banana-odoo) [reply addons/test_mail/tests/test_mail_composer.py]: Yes, forgot we can do it :)

**Review by tde-banana-odoo — COMMENTED** [review]
Updated / answered (runbot not building, github seems to pedale in the semoule, encore).

**Comment by tde-banana-odoo** [comment]
@std-odoo I think everything is answered or updated.

**Comment by std-odoo** [comment]
@tde-banana-odoo LGTM :)

**Comment by tde-banana-odoo** [comment]
Thanks for the review !

**Comment by tde-banana-odoo** [comment]
@robodoo r+ rebase-merge

**Pull request closed without merging**